Feature/admin course management

tobias-wiese opened the pull request:
# Description

Implements issue #438 
Implements issue #439 
-> implements epic #200 

## Reason for this PR
- implement the epic

## Changes in this PR
- add course list and course form to the navbars of the admin
- adapt routing as list and form are now shared between admin and lecturer ("/all-courses")
- enable editing of all courses for the admin
- add a lecturer selection section to the course form that is visible just for admins


## Type of change (remove all that don't apply)
- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Manually tested all use cases this could alter

**Test Configuration**:
- OS: Windows
- Browser: Firefox

- Frontend: (remove all that don't apply)
  - [x] Development build
- Backend: (remove all that don't apply)
  - [x] deployed development build

# Checklist: (remove all that don't apply)

- [x] I added corresponding E2E tests (especially for bugfixes)
- [x] I have performed a self-review of my own code
- [x] My changes generate no new linting warnings or console warnings
- [x] I have updated the CHANGELOG.md to include any changes made in this PR (add WIP to the top, if there is none already)

## Notes:
creating a course with an invalid lecturer Id "works" atm as we do not get a validation error. The validation error is implemented in the backend, but not yet merged. The error handling for that is implemented.

bastihav reviewed (changes requested):
Feature is completely unusable on both chrome and firefox, did you forget to emit the lecturerId?

Whats the difference between the shared/ and common/ folder? Should we merge them?
I am not sure about the "lecturer found" text, at least the styling. I think a more subtle green checkmark or red cross + black text would be more suitible, maybe add a link to the profile of the lecturer too.

E2E tests are planned as soon as the test branch is merged, I guess?

tobias-wiese commented:
> 
> 
> Feature is completely unusable on both chrome and firefox, did you forget to emit the lecturerId?

Yes, that was the problem. Should be fixed now.

> 
> Whats the difference between the shared/ and common/ folder? Should we merge them?

I added the "shared" folder for views, that are shown for multiple roles (here Lecturer and Admin), but not for all roles (as for example the profile pages). It should not be a problem to change merge the folders, I just thought this could be "cleaner".

> I am not sure about the "lecturer found" text, at least the styling. I think a more subtle green checkmark or red cross + black text would be more suitible, maybe add a link to the profile of the lecturer too.

I implement this like in the following screenshots
![grafik](https://user-images.githubusercontent.com/63233799/92414261-cf41d880-f153-11ea-8a45-27473ae87f96.png)
(This is the hover style for the link)
![grafik](https://user-images.githubusercontent.com/63233799/92414289-e7b1f300-f153-11ea-85ef-4d13adaf3d44.png)

> 
> E2E tests are planned as soon as the test branch is merged, I guess?

Yes. I already added a very small test for the new navigation, but the functionality will be tested, as soon the test branch is merged.

bastihav commented on src/views/shared/EditCreateCourseForm.vue:
```diff
@@ -87,7 +87,7 @@
 <script lang="ts">
     import Router from "@/use/router/";
-    import { useStore } from "@/use/store/store";
+    import { store, useStore } from "@/use/store/store";
```
```suggestion
    import { useStore } from "@/use/store/store";
```

did not test if this breaks things, but we should only access the store with getters anyway

bastihav reviewed (commented):
will approve after e2e tests

bastihav commented on tests/e2e/adminCourseManagement.spec.ts:
```diff
@@ -0,0 +1,121 @@
+
+    before(function () {
+        cy.fixture("course.json").then((c) => {
+            course = { ...(c as Course) };
```
```suggestion
            course = c as Course;
```

theovier commented:
I would prefer if the select menu for the lecturers (first image) would look more like the other selection menus we used (second image) just for the sake of consistency. [Chrome on MacOS used].
![image](https://user-images.githubusercontent.com/13664498/92494309-d91c1800-f1f5-11ea-854c-4284746f295c.png)
![image](https://user-images.githubusercontent.com/13664498/92494501-141e4b80-f1f6-11ea-800d-7bb8a83b27dd.png)

Besides that, it looks good and functional 👍

tobias-wiese commented:
> 
> 
> I would prefer if the select menu for the lecturers (first image) would look more like the other selection menus we used (second image) just for the sake of consistency. [Chrome on MacOS used].
> ![image](https://user-images.githubusercontent.com/13664498/92494309-d91c1800-f1f5-11ea-854c-4284746f295c.png)
> ![image](https://user-images.githubusercontent.com/13664498/92494501-141e4b80-f1f6-11ea-800d-7bb8a83b27dd.png)
> 
> Besides that, it looks good and functional 👍

That's because it is a datalist connected to an input field, not a select field. I will try to style the datalist as well

theovier commented:
> > I would prefer if the select menu for the lecturers (first image) would look more like the other selection menus we used (second image) just for the sake of consistency. [Chrome on MacOS used].
> > ![image](https://user-images.githubusercontent.com/13664498/92494309-d91c1800-f1f5-11ea-854c-4284746f295c.png)
> > ![image](https://user-images.githubusercontent.com/13664498/92494501-141e4b80-f1f6-11ea-800d-7bb8a83b27dd.png)
> > Besides that, it looks good and functional 👍
> 
> That's because it is a datalist connected to an input field, not a select field. I will try to style the datalist as well

Just curious, what is the reason that you used a datalist rather than a prefilled select field (like the countries)?

tobias-wiese commented:
> 
> 
> > > I would prefer if the select menu for the lecturers (first image) would look more like the other selection menus we used (second image) just for the sake of consistency. [Chrome on MacOS used].
> > > ![image](https://user-images.githubusercontent.com/13664498/92494309-d91c1800-f1f5-11ea-854c-4284746f295c.png)
> > > ![image](https://user-images.githubusercontent.com/13664498/92494501-141e4b80-f1f6-11ea-800d-7bb8a83b27dd.png)
> > > Besides that, it looks good and functional 👍
> > 
> > 
> > That's because it is a datalist connected to an input field, not a select field. I will try to style the datalist as well
> 
> Just curious, what is the reason that you used a datalist rather than a prefilled select field (like for the languages)?

To enable a search. If we have like 100 lecturers, I want to filter the datalist with the input given in the text input field

theovier commented:
> To enable a search. If we have like 100 lecturers, I want to filter the datalist with the input given in the text input field
You can have a search also with a custom select if I'm not mistaken. As I understood datalist they represent more of `suggestion` than actual choices. However, we require the user to select one of several predefined choices. However, I'm fine with it if the styling matches.

tobias-wiese commented:
> 
> 
> > To enable a search. If we have like 100 lecturers, I want to filter the datalist with the input given in the text input field
> 
> You can have a search also with a custom select if I'm not mistaken. As I understood, datalist represent more of `suggestion` than actual choices. However, we require the user to select one of several predefined choices. However, I'm fine with it if the styling matches.

I do know that implementing such a select is possible. I decided to use the data list as it is a plain HTML5 solution. Otherwise I'd had to implement it using JS (with an input field and different divs for the options)  or use a plugin, but you know the problems with plugins 😄

tobias-wiese commented:
> 
> 
> > > To enable a search. If we have like 100 lecturers, I want to filter the datalist with the input given in the text input field
> > 
> > 
> > You can have a search also with a custom select if I'm not mistaken. As I understood, datalist represent more of `suggestion` than actual choices. However, we require the user to select one of several predefined choices. However, I'm fine with it if the styling matches.
> 
> I do know that implementing such a select is possible. I decided to use the data list as it is a plain HTML5 solution. Otherwise I'd had to implement it using JS (with an input field and different divs for the options) or use a plugin, but you know the problems with plugins 😄

Now I implemented an own select component with an input and absolute divs that looks like our form selects (see screenshots). Styling the datalist was not possible as it is a browser specific component. @bastihav  You may want to look into the implementation again, as I changed and added many things
![grafik](https://user-images.githubusercontent.com/63233799/92585613-7bd7a380-f295-11ea-9189-65e4216e21fb.png)

theovier reviewed (approved):
Looks nice, ty 😘

bastihav commented on src/views/shared/EditCreateCourseForm.vue:
```diff
@@ -126,7 +128,7 @@
         emits: ["update:has-input", "update:success"],
 
-        setup(props: any, { emit }: any) {
+        setup(props: any, { emit, root }: any) {
```
```suggestion
        setup(props: any, { emit }: any) {
```
not used

bastihav commented on tests/e2e/adminCourseManagement.spec.ts:
```diff
@@ -24,12 +28,24 @@ describe("Course creation, edition and deletion", () => {
         cy.fixture("logins/admin.json").then((lecturer) => {
```
```suggestion
        cy.fixture("logins/lecturer.json").then((lecturer) => {
```

bastihav commented on tests/e2e/adminCourseManagement.spec.ts:
```diff
@@ -89,6 +105,21 @@ describe("Course creation, edition and deletion", () => {
     });
 
+    it("Filtering lecturer list in selection works", () => {
+        cy.get("input[id='lecturerId']").clear().type("lecturer");
```
```suggestion
        cy.get("input[id='lecturerId']").clear().type(lecturerAuth.username);
```

bastihav commented on tests/e2e/adminCourseManagement.spec.ts:
```diff
@@ -89,6 +105,21 @@ describe("Course creation, edition and deletion", () => {
 
+    it("Filtering lecturer list in selection works", () => {
+        cy.get("input[id='lecturerId']").clear().type("lecturer");
+        cy.get("div").contains("(@lecturer)").should("exist");
```
```suggestion
        cy.get("div").contains(`(@${lecturerAuth.username})`).should("exist");
```

bastihav commented on tests/e2e/adminCourseManagement.spec.ts:
```diff
@@ -89,6 +105,21 @@ describe("Course creation, edition and deletion", () => {
+    it("Can select default lecturer", () => {
+        cy.get("input[id='lecturerId']").clear();
+        cy.get("input[id='lecturerId']").click();
+        cy.get("div").contains("(@lecturer)").click();
+        cy.get("input[id='lecturerId']").should("have.value", "lecturer");
```
```suggestion
        cy.get("div").contains(`(@${lecturerAuth.username})`).click();
        cy.get("input[id='lecturerId']").should("have.value", lecturerAuth.username);
```

bastihav reviewed (changes requested):
did not test any of the changes, if the tests pass, feel free to merge
